File: android/src/main/java/io/ionic/plugins/applanguage/AppLanguagePlugin.java
package io.ionic.plugins.applanguage;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "AppLanguage")
public class AppLanguagePlugin extends Plugin {

    public static final String TAG = "AppLanguage";

    private AppLanguage implementation;

    public void load() {
        implementation = new AppLanguage(this);
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        try {
            implementation.initialize();
            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void getApplicationLocales(PluginCall call) {
        try {
            JSObject result = new JSObject();

            String[] locales = implementation.getApplicationLocales().split(",");

            if (locales.length > 0) {
                JSArray jsLocales = new JSArray();

                for (String locale : locales) {
                    jsLocales.put(locale);
                }

                result.put("locales", jsLocales);
            }

            call.resolve(result);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void setApplicationLocales(PluginCall call) {
        try {
            JSArray jsLocales = call.getArray("locales");

            if (jsLocales != null && jsLocales.length() > 0) {
                StringBuilder locales = new StringBuilder();

                for (Object locale : jsLocales.toList()) {
                    locales.append(locale);
                }

                implementation.setApplicationLocales(locales.toString());
            }

            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void resetApplicationLocales(PluginCall call) {
        try {
            implementation.resetApplicationLocales();

            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void getSystemLocales(PluginCall call) {
        try {
            JSObject result = new JSObject();

            String[] locales = implementation.getSystemLocales().split(",");

            if (locales.length > 0) {
                JSArray jsLocales = new JSArray();

                for (String locale : locales) {
                    jsLocales.put(locale);
                }

                result.put("locales", jsLocales);
            }

            call.resolve(result);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void getOverrideLocaleConfig(PluginCall call) {
        try {
            JSObject result = new JSObject();

            AppLocaleConfig config = implementation.getOverrideLocaleConfig();

            if (config != null) {
                JSArray jsLocales = new JSArray();

                String[] locales = config.getLocales().split(",");

                for (String locale : locales) {
                    jsLocales.put(locale);
                }

                result.put("locales", jsLocales);
                result.put("status", config.getStatus());
            }

            call.resolve(result);
        } catch (UnsupportedOperationException exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.unavailable(exception.getMessage());
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void setOverrideLocaleConfig(PluginCall call) {
        try {
            JSArray jsLocales = call.getArray("locales");

            if (jsLocales.length() > 0) {
                StringBuilder locales = new StringBuilder();

                for (Object locale : jsLocales.toList()) {
                    locales.append(locale);
                }

                implementation.setOverrideLocaleConfig(locales.toString());
            }

            call.resolve();
        } catch (UnsupportedOperationException exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.unavailable(exception.getMessage());
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void openSettings(PluginCall call) {
        try {
            implementation.openSettings();

            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }
}

File: ios/Sources/AppLanguagePlugin/AppLanguagePlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(AppLanguagePlugin)
public class AppLanguagePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AppLanguagePlugin"
    public let jsName = "AppLanguage"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getApplicationLocales", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setApplicationLocales", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resetApplicationLocales", returnType: CAPPluginReturnPromise),

        CAPPluginMethod(name: "getSystemLocales", returnType: CAPPluginReturnPromise),

        CAPPluginMethod(name: "getOverrideLocaleConfig", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setOverrideLocaleConfig", returnType: CAPPluginReturnPromise),

        CAPPluginMethod(name: "openSettings", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = AppLanguage()

    @objc func getApplicationLocales(_ call: CAPPluginCall) {
        call.resolve([
            "locales": implementation.getApplicationLocales()
        ])
    }

    @objc func setApplicationLocales(_ call: CAPPluginCall) {
        call.unimplemented("Not implemented on iOS.")
    }

    @objc func resetApplicationLocales(_ call: CAPPluginCall) {
        call.unimplemented("Not implemented on iOS.")
    }

    @objc func getSystemLocales(_ call: CAPPluginCall) {
        call.resolve([
            "locales": implementation.getSystemLocales()
        ])
    }

    @objc func getOverrideLocaleConfig(_ call: CAPPluginCall) {
        call.unimplemented("Not implemented on iOS.")
    }

    @objc func setOverrideLocaleConfig(_ call: CAPPluginCall) {
        call.unimplemented("Not implemented on iOS.")
    }

    @objc func openSettings(_ call: CAPPluginCall) {
        implementation.openSettings()

        call.resolve()
    }
}
